Examples/file-manager.lisp

```diff
@@ -0,0 +1,349 @@
+
+(define-presentation-method present ((object root)
+                                     (type   root)
+                                     (stream extended-output-stream)
+                                     (view   file-manager-view)
+                                     &key)
+  (with-bounding-rectangle* (x1 y1 x2 y2) (sheet-region stream)
+    (draw-rectangle* stream (+ x1 4) (+ y1 4) (- x2 4) (- y2 4) :filled nil))
+  (stream-increment-cursor-position stream 8 8)
+  (print-directory-contents (children object) stream))
+
+(define-presentation-method present ((object trash)
+                                     (type   trash)
+                                     (stream extended-output-stream)
+                                     (view   file-manager-view)
+                                     &key)
+  (surrounding-output-with-border (stream :shape :rectangle :background +beige+)
+    (princ (name object) stream)
+    (with-translation (stream (+ (stream-cursor-position stream) 8) 0)
+      (draw-polygon* stream '(0 4 4 2 12 2 16 4 12 4 12 16 4 16 4 4)))))
+
+;;; Application
+
+(defvar *explanation*
+  (format nil "Drag and drop files and directories to move, copy, ~
+               etc.~@
+               The modifier keys change which action is performed when ~
+               the object is dropped:~@
+               • no modifier → the item is moved~@
+               • meta (often alt key) → the item is copied~@
+               Moving things into the \"Trash\" directory deletes them.~@
+               Be careful with recursive directory copies."))
+
+(define-application-frame file-manager ()
+  ((%root :initarg  :root
+          :reader   root
+          :initform (make-root
+                     (make-instance 'trash)
+                     (labels ((rec (class &optional (name (class-name class)))
+                                (if-let ((children (c2mop:class-direct-subclasses class)))
+                                  (apply #'make-directory name (mapcar #'rec children))
+                                  (make-file name))))
+                       (rec (find-class 'number) "Number types"))
+                     (apply #'make-directory "Features"
+                            (mapcar #'make-file
+                                    (subseq *features* 0 (min (length *features*) 10)))))))
+  (:panes
+   (explananation :label :label      *explanation*
+                         :background +beige+)
+   (files         :application :display-function (lambda (frame pane)
+                                                   (let ((root (root frame)))
+                                                     (present root (presentation-type-of root)
+                                                              :stream     pane
+                                                              :single-box t)))
+                               :default-view     (make-instance 'file-manager-view))
+   (interactor    :interactor))
```

Comment:
please remove column indent, it looks odd with multiline arguments.

Reply:
Changed.